bin/wasm-node/javascript/src/index-browser.ts

```diff
@@ -121,23 +113,294 @@ function trustedBase64Decode(base64: string): Uint8Array {
       };
 
+      return {
+        close: (): void => {
+            connection.onopen = null;
+            connection.onclose = null;
+            connection.onmessage = null;
+            connection.onerror = null;
+            connection.close();
+        },
+
+        send: (data: Uint8Array): void => {
+            connection.send(data);
+        },
+
+        openOutSubstream: () => { throw new Error('Wrong connection type') }
+      };
+  } else if (webRTCParsed != null) {
+    const targetPort = webRTCParsed[3];
+    if (forbidWebRTC || targetPort === '0') {
+        throw new ConnectionError('Connection type not allowed');
+    }
+
+    const ipVersion = webRTCParsed[1] == 'ip4'? '4' : '6';
+    const targetIp = webRTCParsed[2];
+    const remoteCertMultibase = webRTCParsed[4]!;
+
+    // The payload of `/certhash` is the hash of the self-generated certificate that the
+    // server presents.
+    // This function throws an exception if the certhash isn't correct. For this reason, this call
+    // is performed as part of the parsing of the multiaddr.
+    const remoteCertMultihash = multibaseBase64Decode(remoteCertMultibase);
+    const remoteCertSha256Hash = multihashToSha256(remoteCertMultihash);
+
+    let pc: RTCPeerConnection | null = null;
+    const dataChannels = new Map<number, RTCDataChannel>();
+    // TODO: this system is a complete hack
+    let isFirstSubstream = true;
+    // The opening of the connection is asynchronous. If smoldot calls `close` in the meanwhile,
+    // this variable is set to `true`, and we interrupt the opening.
+    let cancelOpening = false;
+
+    // Function that configures a newly-opened channel and adds it to the map. Used for both
+    // inbound and outbound substreams.
+    const addChannel = (dataChannel: RTCDataChannel, direction: 'inbound' | 'outbound') => {
+      const dataChannelId = dataChannel.id!;
+
+      dataChannel.onopen = () => {
+          config.onStreamOpened(dataChannelId, direction);
+      };
+
+      dataChannel.onerror = (_error) => {
+          config.onStreamClose(dataChannelId);
+      };
+
+      dataChannel.onclose = () => {
+          config.onStreamClose(dataChannelId);
+      };
+
+      dataChannel.onmessage = (m) => {
+          // The `data` field is an `ArrayBuffer`.
+          config.onMessage(new Uint8Array(m.data), dataChannelId);
+      }
+
+      dataChannels.set(dataChannelId, dataChannel);
+    }
+
+    // It is possible for the browser to use multiple different certificates.
+    // In order for our local certificate to be deterministic, we need to generate it manually and
+    // set it explicitly as part of the configuration.
+    // According to <https://w3c.github.io/webrtc-pc/#dom-rtcpeerconnection-generatecertificate>,
+    // browsers are guaranteed to support `{ name: "ECDSA", namedCurve: "P-256" }`.
+    RTCPeerConnection.generateCertificate({ name: "ECDSA", namedCurve: "P-256" } as EcKeyGenParams).then((localCertificate) => {
```

Comment:
I think we can specify a hash algorithm here. See https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/generateCertificate#specifying_algorithm_details

that way we won't need the for loop below